src/common/re2_regex.cpp

```diff
@@ -0,0 +1,90 @@
+#include "common/re2_regex.h"
+
+#include <regex>
+
+namespace kuzu {
+namespace common {
+
+Regex::Regex(const std::string& pattern, RegexOptions options) {
+    regex::RE2::Options o;
+    o.set_case_sensitive(options == RegexOptions::CASE_INSENSITIVE);
+    regex = std::make_shared<regex::RE2>(regex::StringPiece(pattern), o);
+}
+
+bool RegexSearchInternal(const char* input, Match& match, const Regex& r, regex::RE2::Anchor anchor,
+    size_t start, size_t end) {
+    auto& regex = r.getRegex();
+    std::vector<regex::StringPiece> targetSubMatches;
+    auto submatchCount = regex.NumberOfCapturingGroups() + 1;
+    targetSubMatches.resize(submatchCount);
+    match.clear();
+    if (!regex.Match(regex::StringPiece(input), start, end, anchor, targetSubMatches.data(),
+            submatchCount)) {
+        return false;
+    }
+    for (auto& subMatch : targetSubMatches) {
+        SubMatch group_match(subMatch.ToString(), subMatch.data() - input);
+        match.emplace_back(group_match);
+    }
+    return true;
+}
+
+bool RegexFullMatch(const std::string& input, const Regex& regex) {
+    Match nopMatch;
+    return RegexSearchInternal(
+        input.c_str(), nopMatch, regex, regex::RE2::ANCHOR_BOTH, 0, input.size());
+}
+
+bool RegexPartialMatch(const std::string& input, const Regex& regex) {
+    Match nopMatch;
+    return RegexSearchInternal(
+        input.c_str(), nopMatch, regex, regex::RE2::UNANCHORED, 0, input.size());
+}
+
+std::string RegexReplace(
+    const std::string& input, const Regex& regex, const std::string& replacement) {
+    std::string result = input;
+    RE2::Replace(&result, regex.getRegex(), replacement);
+    return result;
+}
+
+bool RegexExtract(const std::string& input, Match& match, const Regex& regex) {
+    return RegexSearchInternal(
+        input.c_str(), match, regex, regex::RE2::UNANCHORED, 0, input.size());
+}
+
+static inline bool IsCharacter(char c) {
+    // Check if this character is not the middle of utf-8 character i.e. it shouldn't begin with 10
+    // XX XX XX
+    return (c & 0xc0) != 0x80;
+}
+
+std::vector<Match> RegexExtractAll(const std::string& input, const Regex& regex) {
+    std::vector<Match> matches;
+    size_t position = 0;
+    Match match;
+    while (RegexSearchInternal(
+        input.c_str(), match, regex, regex::RE2::UNANCHORED, position, input.size())) {
+        size_t newPosition = match.getPosition(0) + match.length(0);
+        if (newPosition == position) {
+            // Empty match found, increment the position manually
+            newPosition++;
+            while (newPosition < input.length() && !IsCharacter(input[newPosition])) {
+                newPosition++;
+            }
+        }
+        position = newPosition;
+        matches.emplace_back(std::move(match));
+    }
+    return matches;
+}
+
+std::string ParseCypherRegex(const std::string& input) {
```

Comment:
We use camel case for function/variable names. Rename this function to `parseCypherRegex`